classad/_grammar.py

```diff
@@ -0,0 +1,232 @@
+import pyparsing as pp
+
+from classad._base_expression import Expression
+from classad._expression import (
+    ClassAd,
+    AttributeExpression,
+    FunctionExpression,
+    ArithmeticExpression,
+    SubscriptableExpression,
+    TernaryExpression,
+    NamedExpression,
+    UnaryExpression,
+)
+from classad._primitives import (
+    Error,
+    Undefined,
+    HTCBool,
+    HTCInt,
+    HTCFloat,
+    HTCStr,
+    HTCList,
+)
+
+pp.ParserElement.enablePackrat()
+
+SQUOTE = pp.Suppress("'")
+DQUOTE = pp.Suppress('"')
+LPAR = pp.Suppress("(")
+RPAR = pp.Suppress(")")
+LBRACKET = pp.Suppress("[")
+RBRACKET = pp.Suppress("]")
+LBRACE = pp.Suppress("{")
+RBRACE = pp.Suppress("}")
+
+# Reserved Words
+boolean_literal = pp.oneOf("true false", caseless=True, asKeyword=True).setParseAction(
+    lambda s, l, t: HTCBool(1) if t[0] in ["true"] else HTCBool(0)
+)
+error_literal = pp.CaselessKeyword("error").setParseAction(lambda: Error())
+undefined_literal = pp.CaselessKeyword("undefined").setParseAction(lambda: Undefined())
+parent_literal = pp.CaselessKeyword("parent").setParseAction(
+    lambda s, l, t: NamedExpression.from_grammar(t[0])
+)
+target_literal = pp.CaselessKeyword("target").setParseAction(
+    lambda s, l, t: NamedExpression.from_grammar(t[0])
+)
+super_literal = pp.CaselessKeyword("super").setParseAction(
+    lambda s, l, t: NamedExpression.from_grammar(t[0])
+)
+
+# Tokens
+octal_digit = pp.srange("[0-7]")
+nonzero_digit = pp.srange("[1-9]")
+decimal_digit = pp.nums
+reserved_word = (
```

Comment:
If you have performance problems, consider using ``MatchFirst((error_literal, boolean_literal, ...))`` to flatten the match expression.

Reply:
I've done so. However, it does not seem to increase performance a lot or even at all ;)